components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/ServiceDiscoverer.java

```diff
@@ -0,0 +1,151 @@
+     * @throws ServiceDiscoveryException if an error occurs in the implementation's initImpl method
+     */
+    public void init(HashMap<String, String> implParameters) throws ServiceDiscoveryException {
+        this.namespaceFilter = implParameters.get(ServiceDiscoveryConstants.NAMESPACE);
+
+        // Convert the criteria that is passed in as a string into a map and then set the instance criteriaFilter
+        String criteriaString = implParameters.get(ServiceDiscoveryConstants.CRITERIA);
+        if (criteriaString != null) {
+            String[] criteriaArray = criteriaString.split(",");
+            HashMap<String, String> criteriaMap = new HashMap<>();
+            for (String pair : criteriaArray) {
+                String[] entry = pair.split("=");
+                criteriaMap.put(entry[0].trim(), entry[1].trim());
+            }
+            this.criteriaFilter = criteriaMap;
+        }
+        serviceEndpointIndex = 0;
+        initImpl(implParameters);
+    }
+
+    /**
+     * Initialization method for all implementation classes
+     *
+     * @param implParameters implementation parameters added by #init(java.util.HashMap) method
+     * @throws ServiceDiscoveryException if an error occurs while initializing
+     */
+    abstract void initImpl(HashMap<String, String> implParameters) throws ServiceDiscoveryException;
+
+    /**
+     * Gives a list of endpoints without any filtering.
+     *
+     * @return List of all Endpoints
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices() throws ServiceDiscoveryException;
+
+    /**
+     * Gives a list of endpoints filtered by namespace.
+     *
+     * @param namespace     Namespace of the expected endpoints
+     * @return List of Endpoints with the specified namespace
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices(String namespace) throws ServiceDiscoveryException;
+
+    /**
+     * Gives a list of endpoints filtered by criteria.
+     *
+     * @param criteria    A set of criteria which all the endpoints must belong to
+     * @return List of Endpoints with the specified criteria
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices(HashMap<String, String> criteria) throws ServiceDiscoveryException;
+
+    /**
+     * Gives a list of endpoints filtered by both namespace and criteria.
+     *
+     * @param namespace   Namespace of the expected endpoints
+     * @param criteria    A set of criteria which all the endpoints must belong to
+     * @return List of Endpoints with the specified namespace and criteria
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices(String namespace, HashMap<String, String> criteria)
```

Comment:
Is there a specific reason to use use HashMap in the method signature? if not use Map instead of HashMap